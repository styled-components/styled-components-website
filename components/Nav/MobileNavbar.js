import React from 'react';
import styled, { css } from 'styled-components';
import { Search, KeyboardArrowDown } from '@styled-icons/material';
import rem from '../../utils/rem';
import { navbarHeight } from '../../utils/sizes';
import { paleGrey } from '../../utils/colors';
import { mobile } from '../../utils/media';
import { CloseIcon, FoldIcon } from './NavIcons';
import Link from '../Link';
import NavLinks from './NavLinks';
import Social from './Social';
import Logo from './Logo';
import NavButton from './NavButton';

const Wrapper = styled.div`
  display: none;

  ${mobile(css`
    display: flex;
    align-items: center;
    justify-content: space-between;
    height: ${rem(navbarHeight)};
  `)};
`;

const SecondaryMenu = styled.div`
  position: absolute;
  top: ${rem(navbarHeight)};
  left: 0;
  right: 0;

  ${p =>
    p.isOpen
      ? css`
          height: ${rem(navbarHeight)};
        `
      : css`
          height: 0;
        `} display: flex;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: space-between;
  padding: 0 ${rem(20)};
  transition: height 0.1s;

  user-select: none;
  -webkit-overflow-scrolling: touch;
  overflow-x: scroll;
  overflow-y: hidden;

  background: ${paleGrey};
  color: #868686;
`;

const LogoLink = styled(Link).attrs((/* props */) => ({
  unstyled: true,
  href: '/',
  'aria-label': 'styled components',
}))`
  display: inline-block;
  vertical-align: center;
`;

const ArrowWrapper = styled.div`
  transition: transform 0.2s;

  ${p =>
    p.shouldRotate &&
    css`
      transform-origin: center center;
      transform: rotate(180deg);
    `};
`;

const SecondaryMenuItem = styled.div`
  padding-right: ${rem(20)};
`;

const StyledIcon = styled.div`
  && {
    width: ${p => rem(p.size || 20)};
    height: ${p => rem(p.size || 20)};
  }
`;

const MobileNavbar = props => {
  const { isSideFolded, isMobileNavFolded, onSideToggle, onMobileNavToggle, showSideNav, onSearchButtonClick } = props;

  return (
    <Wrapper>
      {showSideNav !== false && (
        <NavButton active={!isSideFolded} onClick={onSideToggle}>
          {isSideFolded ? <FoldIcon /> : <CloseIcon />}
        </NavButton>
      )}

      <LogoLink>
        <Logo compact />
      </LogoLink>
      <div>
        <NavButton onClick={onSearchButtonClick}>
          <StyledIcon as={Search} size={28} />
        </NavButton>

        <NavButton onClick={onMobileNavToggle} active={!isMobileNavFolded}>
          <ArrowWrapper shouldRotate={!isMobileNavFolded}>
            <StyledIcon as={KeyboardArrowDown} size={36} />
          </ArrowWrapper>
        </NavButton>
      </div>

      <SecondaryMenu isOpen={!isMobileNavFolded}>
        <NavLinks />
        <SecondaryMenuItem>
          <Social />
        </SecondaryMenuItem>
      </SecondaryMenu>
    </Wrapper>
  );
};

export default MobileNavbar;
