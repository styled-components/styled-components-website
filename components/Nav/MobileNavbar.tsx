import { KeyboardArrowDown } from '@styled-icons/material';
import styled, { css } from 'styled-components';
import { paleGrey } from '../../utils/colors';
import { mobile } from '../../utils/media';
import rem from '../../utils/rem';
import { navbarHeight } from '../../utils/sizes';
import Link from '../Link';
import Logo from './Logo';
import NavButton from './NavButton';
import { CloseIcon, FoldIcon } from './NavIcons';
import NavLinks from './NavLinks';
import Social from './Social';

const Wrapper = styled.div`
  display: none;

  ${mobile(css`
    align-items: center;
    display: flex;
    flex-grow: 1;
    height: ${rem(navbarHeight)};
    justify-content: space-between;
  `)};
`;

const SecondaryMenu = styled.div<{ $isOpen?: boolean }>`
  position: absolute;
  top: ${rem(navbarHeight)};
  left: 0;
  right: 0;

  ${p =>
    p.$isOpen
      ? css`
          height: ${rem(navbarHeight)};
        `
      : css`
          height: 0;
        `} display: flex;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: space-between;
  padding: 0 ${rem(20)};
  transition: height 0.1s;

  user-select: none;
  -webkit-overflow-scrolling: touch;
  overflow-x: scroll;
  overflow-y: hidden;

  background: ${paleGrey};
  color: #868686;
`;

const LogoLink = styled(Link).attrs((/* props */) => ({
  unstyled: true,
  href: '/',
  'aria-label': 'styled components',
}))`
  display: inline-block;
  vertical-align: center;
`;

const ArrowWrapper = styled.div<{ $shouldRotate?: boolean }>`
  transition: transform 0.2s;

  ${p =>
    p.$shouldRotate &&
    css`
      transform-origin: center center;
      transform: rotate(180deg);
    `};
`;

const SecondaryMenuItem = styled.div`
  padding-right: ${rem(20)};
`;

const StyledIcon = styled.div<{ $size?: number }>`
  && {
    width: ${p => rem(p.$size || 20)};
    height: ${p => rem(p.$size || 20)};
  }
`;

export interface MobileNavbarProps {
  isMobileNavFolded?: boolean;
  isSideFolded?: boolean;
  onMobileNavToggle?: () => void;
  onSideToggle?: () => void;
  showSideNav?: boolean;
}

const MobileNavbar = ({
  children,
  isMobileNavFolded,
  isSideFolded,
  onMobileNavToggle,
  onSideToggle,
  showSideNav,
}: React.PropsWithChildren<MobileNavbarProps>) => {
  return (
    <Wrapper>
      {showSideNav !== false && (
        <NavButton $active={!isSideFolded} onClick={onSideToggle}>
          {isSideFolded ? <FoldIcon /> : <CloseIcon />}
        </NavButton>
      )}

      <LogoLink>
        <Logo $compact />
      </LogoLink>

      {children}

      <NavButton onClick={onMobileNavToggle} $active={!isMobileNavFolded} style={{ position: 'absolute', right: 0 }}>
        <ArrowWrapper $shouldRotate={!isMobileNavFolded}>
          <StyledIcon as={KeyboardArrowDown} $size={36} />
        </ArrowWrapper>
      </NavButton>

      <SecondaryMenu $isOpen={!isMobileNavFolded}>
        <NavLinks />
        <SecondaryMenuItem>
          <Social />
        </SecondaryMenuItem>
      </SecondaryMenu>
    </Wrapper>
  );
};

export default MobileNavbar;
