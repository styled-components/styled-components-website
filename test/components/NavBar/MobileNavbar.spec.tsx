import React from 'react';
import renderer from 'react-test-renderer';
import MobileNavbar from '../../../components/Nav/MobileNavbar';

test('MobileNavbar renders correctly', () => {
  const tree = renderer.create(<MobileNavbar />).toJSON();

  expect(tree).toMatchSnapshot();
});
