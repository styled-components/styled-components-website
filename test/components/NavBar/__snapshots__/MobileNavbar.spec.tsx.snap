// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileNavbar renders correctly 1`] = `
.c5 {
  background-image: url(/nav-logo.png);
  background-position: center;
  background-size: contain;
  box-sizing: border-box;
  display: inline-block;
  height: 2.2222222222222223rem;
  width: 9.11111111111111rem;
  vertical-align: middle;
}

.c2 {
  background: none;
  outline: none;
  border: none;
  background: transparent;
  cursor: pointer;
  flex: 0 0 auto;
  height: 3.3333333333333335rem;
  padding: 0 0.5555555555555556rem;
  text-align: center;
  transition: all 200ms ease-in-out;
  vertical-align: middle;
}

.c2:hover {
  filter: brightness(0.85);
}

.c2:active {
  filter: brightness(1);
}

.c1+.c1 {
  padding-left: 0;
}

.c3 svg {
  display: inline-block;
}

.c3 svg path {
  fill: currentColor;
}

.c11 {
  flex: 0 0 auto;
  width: 0.2777777777777778rem;
  height: 0.2777777777777778rem;
  margin: 0 0.8333333333333334rem;
  border-radius: 50%;
  background: currentColor;
  opacity: 0.35;
}

.c9 {
  display: flex;
  align-items: center;
  flex: 0 0 auto;
  margin-right: 1.6666666666666667rem;
}

.c10 {
  flex: 0 0 auto;
  display: inline-block;
  line-height: 3.3333333333333335rem;
  transition: opacity 0.2s,transform 0.2s;
  cursor: pointer;
  letter-spacing: 0.022222222222222223rem;
  color: currentColor;
}

.c10:hover,
.c10:focus {
  opacity: 0.8;
}

.c10:active {
  transform: scale(0.95);
  opacity: 0.6;
}

.c0 {
  display: none;
}

.c8 {
  position: absolute;
  top: 3.3333333333333335rem;
  left: 0;
  right: 0;
  height: 3.3333333333333335rem;
  display: flex;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: space-between;
  padding: 0 1.1111111111111112rem;
  transition: height 0.1s;
  user-select: none;
  -webkit-overflow-scrolling: touch;
  overflow-x: scroll;
  overflow-y: hidden;
  background: rgb(33, 33, 33);
  color: #fff;
}

.c4 {
  margin-right: auto;
  transform: translateY(-1px);
}

.c6 {
  transition: transform 0.2s;
  transform-origin: center center;
  transform: rotate(180deg);
}

.c7.c7 {
  width: 2rem;
  height: 2rem;
}

@media (max-width: 62.5em) {
  .c0 {
    align-items: center;
    display: flex;
    flex-grow: 1;
    height: 3.3333333333333335rem;
    justify-content: space-between;
  }
}

<div
  className="c0"
>
  <button
    className="c1 c2"
  >
    <svg
      className="c3"
      height="15"
      width="15"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <title>
        close
      </title>
      <use
        fill="#FFF"
        transform="translate(1 1)"
        xlinkHref="#close"
      />
      <defs>
        <path
          d="M-.7.7l13 13 1.4-1.4-13-13L-.7.7zm13-1.4l-13 13 1.4 1.4 13-13-1.4-1.4z"
          id="close"
        />
      </defs>
    </svg>
  </button>
  <a
    aria-label="styled components"
    className="c4"
  >
    <div
      className="c5"
    />
  </a>
  <button
    className="c1 c2"
    style={
      {
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <div
      className="c6"
    >
      <div
        className="c7"
      />
    </div>
  </button>
  <div
    className="c8"
  >
    <nav
      className="c9"
    >
      <a
        className="c10"
      >
        Documentation
      </a>
      <span
        className="c11"
      />
      <a
        className="c10"
      >
        Showcase
      </a>
      <span
        className="c11"
      />
      <a
        className="c10"
      >
        Ecosystem
      </a>
      <span
        className="c11"
      />
      <a
        className="c10"
      >
        Releases
      </a>
    </nav>
  </div>
</div>
`;
